src/main/java/com/aws/greengrass/secretmanager/SecretManager.java

```diff
@@ -69,74 +78,78 @@ public class SecretManager {
     private final AWSSecretClient secretClient;
     private final SecretStore<SecretDocument, AWSSecretResponse> secretStore;
     @Nullable
-    private Crypter crypter;
-    private final Result<SecretCryptoException> initialized = new Result<>();
+    private final AtomicReference<Crypter> crypter = new AtomicReference<>();
 
     /**
      * Constructor.
-     * @param secretClient client for aws secrets.
+     *
+     * @param secretClient    client for aws secrets.
      * @param securityService security service.
-     * @param dao          dao for persistent store.
+     * @param dao             dao for persistent store.
      */
     @Inject
     SecretManager(AWSSecretClient secretClient, SecurityService securityService, FileSecretStore dao,
-                  ExecutorService executor) {
+                  DeviceConfiguration deviceConfiguration) {
         this.secretStore = dao;
         this.secretClient = secretClient;
+        this.securityService = securityService;
+        deviceConfiguration.onAnyChange(((whatHappened, node) -> {
+            if (validUpdate(node, DEVICE_PARAM_CERTIFICATE_FILE_PATH) || validUpdate(node,
+                    DEVICE_PARAM_PRIVATE_KEY_PATH)) {
+                // TODO: If the device creds change, then cache and local store should be updated as well as the keys
+                //  used for encrypting and decrypting the secrets have changed.
+                synchronized (crypter) {
+                    crypter.set(null);
+                }
+            }
+        }));
+    }
+
+    private boolean validUpdate(Node node, String key) {
+        return node != null && node.childOf(key) && Utils.isNotEmpty(Coerce.toString(node)) && Files.exists(
+                Paths.get(key));
+    }
 
-        executor.execute(() -> loadCrypter(securityService));
+    protected Crypter getCrypter() throws SecretCryptoException {
+        synchronized (Objects.requireNonNull(crypter)) {
+            if (crypter.get() == null) {
+                try {
+                    loadCrypter();
```

Comment:
loading the crypter will retry forever. You're going to be blocking the calling thread indefinitely

Reply:
That's right but if the crypto service is not up, only the threads that depend on encryption and decryption are blocked. This will still be the case even when we use a separate thread for loading the crypter. If a thread only needs a secret from cache, it is not blocked by this. 

However, I can change this to retry for ~5 minutes as it shouldn't take longer than that for the services to load.

Reply:
Infinite retry is needed ultimately. I just want you to be sure about what thread is calling this. You must not block critical threads.

Also, use the cycle detecting lock since you depend on nucleus 2.13 anyway.